Python/ceval.c

```diff
@@ -4617,6 +4635,44 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
             DISPATCH();
         }
 
+        TARGET(PRECALL_PY_CLASS) {
+            SpecializedCacheEntry *cache = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &cache[0].adaptive;
+            _PyCallCache *cache1 = &cache[-1].call;
+            int original_oparg = cache->adaptive.original_oparg;
+            int is_method = (PEEK(original_oparg + 2) != NULL);
+            DEOPT_IF(is_method, PRECALL);
+            PyObject *cls = PEEK(original_oparg + 1);
+            DEOPT_IF(!PyType_Check(cls), PRECALL);
+            PyTypeObject *cls_t = (PyTypeObject *)cls;
+            DEOPT_IF(cls_t->tp_version_tag != cache0->version, PRECALL);
+            assert(cls_t->tp_flags & Py_TPFLAGS_HEAPTYPE);
+            PyObject *init = ((PyHeapTypeObject *)cls_t)->_spec_cache.init;
+            assert(PyFunction_Check(init));
+            DEOPT_IF(((PyFunctionObject *)(init))->func_version != cache1->func_version, PRECALL);
+            DEOPT_IF(cls_t->tp_new != PyBaseObject_Type.tp_new, PRECALL);
+            STAT_INC(PRECALL, hit);
+
+            PyObject *args = _PyTuple_FromArray(&PEEK(original_oparg), original_oparg);
+            if (args == NULL) {
+                goto error;
+            }
+            PyObject *self = PyBaseObject_Type.tp_new(cls_t, args, call_shape.kwnames);
```

Comment:
Perhaps we could even call `type_new` here directly, so we don't have to follow a pointer to get the function. Even better, we could expose a slightly different function that takes an array of args directly, so we can skip the tuple conversion.